### PR TITLE
[DFT] portFFT integration

### PR DESCRIPTION
**Pull request by FMarno**
# Description

Adds the portFFT library as a backend for oneMKL.

# Checklist

## All Submissions

- [x] Do all unit tests pass locally? Attach a log.
- [x] Have you formatted the code using clang-format?

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -488,6 +498,8 @@ Python | 3.6 or higher | No | *N/A* | *Pre-installed or Installed by user* | [PS
 [NETLIB LAPACK](https://www.netlib.org/) | 3.7.1 | Yes | conan-community | ~/.conan/data or $CONAN_USER_HOME/.conan/data | [BSD like license](http://www.netlib.org/lapack/LICENSE.txt)
 [Sphinx](https://www.sphinx-doc.org/en/master/) | 2.4.4 | Yes | pip | ~/.local/bin (or similar user local directory) | [BSD License](https://github.com/sphinx-doc/sphinx/blob/3.x/LICENSE)
 [portBLAS](https://github.com/codeplaysoftware/portBLAS) | 0.1 | No | *N/A* | *Installed by user* | [Apache License v2.0](https://github.com/codeplaysoftware/portBLAS/blob/master/LICENSE)
+[portFFT](https://github.com/codeplaysoftware/portFFT) | 0.1 | No | *N/A* | *Installed by user* | [Apache License v2.0](https://github.com/codeplaysoftware/portFFT/blob/master/LICENSE)
```
Comment (hjabird) [inline README.md]: Isn't this downloaded automatically by `fetchContent`?

File: examples/README.md [hunk 2]
```diff
@@ -455,4 +455,29 @@ Device name is: AMD Radeon PRO W6800
...
 DFT example run_time dispatch
 DFT example ran OK
-```
\ No newline at end of file
+```
+
+```none
+$ LD_LIBRARY_PATH=lib ./bin/example_dft_real_fwd_usm 
+########################################################################
+# DFTI complex in-place forward transform with USM API example:
```
Comment (hjabird) [inline examples/README.md]: `DFTI` looks like a hangover from old MKL.

File: examples/README.md [hunk 3]
```diff
@@ -455,4 +455,29 @@ Device name is: AMD Radeon PRO W6800
...
+# SYCL device
+#
+########################################################################
+
+Running DFT complex forward example on GPU device
+Device name is: Intel(R) UHD Graphics 750
+Running with single precision real data type:
+DFT example run_time dispatch
+Caught std::exception:
+	oneMKL: dft/backends/portfft/commit: function is not implemented portFFT only supports complex to complex transforms
```
Comment (hjabird) [inline examples/README.md]: I think we should change the example so something that actually does something:
* Demoing unsupported functionality isn't useful.
* We'll probably forget to update this when the real to complex work is merged.

File: src/dft/backends/portfft/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+		portfft
+		GIT_REPOSITORY https://github.com/codeplaysoftware/portFFT.git
+		GIT_TAG        49f667dca32921747100ec0d79b21940d516a4eb
+	)
+	FetchContent_MakeAvailable(portfft)
+	message(STATUS "portFFT - downloaded")
+	target_link_libraries(${LIB_OBJ} PRIVATE portfft)
+else()
+	message(STATUS "portFFT - found")
+	target_link_libraries(${LIB_OBJ} PRIVATE portfft::portfft)
```
Comment (hjabird) [inline src/dft/backends/portfft/CMakeLists.txt:82]: Why are the link library arguments different?

File: src/dft/backends/portfft/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+target_include_directories(${LIB_OBJ}
+  PRIVATE ${PROJECT_SOURCE_DIR}/include
+          ${PROJECT_SOURCE_DIR}/src
+          ${CMAKE_BINARY_DIR}/bin
+          ${MKL_INCLUDE}
+)
+
+target_compile_options(${LIB_OBJ} PRIVATE ${ONEMKL_BUILD_COPT} ${MKL_COPT})
+
+target_link_libraries(${LIB_OBJ} PUBLIC ONEMKL::SYCL::SYCL ${MKL_LINK_SYCL})
```
Comment (hjabird) [inline src/dft/backends/portfft/CMakeLists.txt]: For those of us who can't remember the cmake for the other backends, is there a distinction here between ONEMKL and MKL?

File: src/dft/backends/portfft/commit.cpp [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+
+#include "portfft_helper.hpp"
+
+// alias to avoid ambiguity
+namespace pfft = portfft;
+
+namespace oneapi::mkl::dft::portfft {
+namespace detail {
+
+/// Commit impl class specialization for cuFFT.
```
Comment (hjabird) [inline src/dft/backends/portfft/commit.cpp]:
```suggestion
/// Commit impl class specialization for portfft.
```

**Review by hjabird — COMMENTED** [review]
Looks pretty good!

File: CMakeLists.txt [hunk 7]
```diff
@@ -51,6 +51,7 @@ endif()
 option(ENABLE_CUBLAS_BACKEND "Enable the cuBLAS backend for the BLAS interface" OFF)
 option(ENABLE_ROCBLAS_BACKEND "Enable the rocBLAS backend for the BLAS interface" OFF)
 option(ENABLE_NETLIB_BACKEND "Enable the Netlib backend for the BLAS interface" OFF)
+option(ENABLE_PORTBLAS_BACKEND "Enable the portBLAS backend for the BLAS interface. Cannot be used with other BLAS backends." OFF)
```
Comment (lhuot) [inline CMakeLists.txt:54]: Is this supposed to be part of this PR?

File: src/dft/backends/portfft/commit.cpp [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+            if (!queue.get_device().has(sycl::aspect::fp64)) {
+                throw mkl::exception("DFT", "commit", "Device does not support double precision.");
+            }
+        }
+    }
+
+    void commit(const dft::detail::dft_values<prec, dom>& config_values) override {
+        // not available in portFFT:
+        if (config_values.workspace != config_value::ALLOW) {
+            throw mkl::unimplemented("dft/backends/portfft", __FUNCTION__,
```
Comment (lhuot) [inline src/dft/backends/portfft/commit.cpp:73]: Maybe it is already done and I missed it but it seems that portFFT support only C2C FFT. So can we make sure to throw an unimplemented  exception if user try to compute R2C or C2R FFT with portFFT backend?

**Comment by lhuot** [comment]
@FMarno could you please attach result of tests to the PR?

File: examples/README.md (hunk 3 above)
Comment (FMarno) [reply examples/README.md]: We discussed this here: https://github.com/FMarno/oneMKL/pull/1#discussion_r1387853267

Do you have a suggestion for what to do?
I don't really want to add a new example. I could potentially not enable the example for portFFT.

File: examples/README.md (hunk 2 above)
Comment (FMarno) [reply examples/README.md]: I always read it as DFT interface.

Comment (hjabird) [reply examples/README.md]: Perhaps thats what it stands for. It appears in the closed-source MKL as -for example- `DFTI_INPLACE`. Maybe its fine if it stands for something sensible then.

File: examples/README.md (hunk 3 above)
Comment (hjabird) [reply examples/README.md]: If its been discussed then there is no point discussing again.

File: src/dft/backends/portfft/CMakeLists.txt (hunk 4 above)
Comment (FMarno) [reply src/dft/backends/portfft/CMakeLists.txt:82]: FetchContent will add the project using add_subdirectory, so the portfft library is added. FindPackage will use the export statement here https://github.com/codeplaysoftware/portFFT/blob/main/CMakeLists.txt#L139, so will put it in a namespace.

File: src/dft/backends/portfft/commit.cpp (hunk 8 above)
Comment (FMarno) [reply src/dft/backends/portfft/commit.cpp:73]: During commit, portFFT will throw a `portfft::unsupported_configuration` exception, which will be caught and `oneapi::mkl::unimplemented` will be thrown with the details.

File: CMakeLists.txt (hunk 7 above)
Comment (FMarno) [reply CMakeLists.txt:54]: This is orthogonal to the PR but I used portblas as inspiration for my cmake and noticed this was missing.

File: src/dft/backends/portfft/CMakeLists.txt (hunk 4 above)
Comment (hjabird) [reply src/dft/backends/portfft/CMakeLists.txt:82]: Makes sense!

File: README.md (hunk 1 above)
Comment (FMarno) [reply README.md]: Yes, I've added a note about that https://github.com/oneapi-src/oneMKL/pull/416/commits/45535b4e8df97f73aa683a677f41d261c7147915

File: src/dft/backends/portfft/CMakeLists.txt (hunk 5 above)
Comment (FMarno) [reply src/dft/backends/portfft/CMakeLists.txt]: MKL is closed source, so not needed. I've removed it here https://github.com/oneapi-src/oneMKL/pull/416/commits/45535b4e8df97f73aa683a677f41d261c7147915

File: src/dft/backends/portfft/commit.cpp (hunk 6 above)
Comment (FMarno) [reply src/dft/backends/portfft/commit.cpp]: Well spotted. It's a pretty useless comment anyway and has been removed. https://github.com/oneapi-src/oneMKL/pull/416/commits/45535b4e8df97f73aa683a677f41d261c7147915

**Comment by FMarno** [comment]
test run on integrated graphics and cpu: 
[portfft_cpu_gpu_run.txt](https://github.com/oneapi-src/oneMKL/files/13478111/portfft_cpu_gpu_run.txt)

**Review by Rbiessy — APPROVED** [review]

**Comment by FMarno** [comment]
[amd_portfft_run.txt](https://github.com/oneapi-src/oneMKL/files/13545830/amd_portfft_run.txt)
[nvidia_portfft_run.txt](https://github.com/oneapi-src/oneMKL/files/13547558/nvidia_portfft_run.txt)

**Comment by FMarno** [comment]
[intel_cpu_gpu_portfft_run.txt](https://github.com/oneapi-src/oneMKL/files/13548995/intel_cpu_gpu_portfft_run.txt)

**Comment by FMarno** [comment]
@lhuot I've uploaded test runs for all the hardware I plan to test.

**Comment by FMarno** [comment]
[amd_2024_run.txt](https://github.com/oneapi-src/oneMKL/files/13583718/amd_2024_run.txt)
[intel_gpu_cpu_2024_run.txt](https://github.com/oneapi-src/oneMKL/files/13583719/intel_gpu_cpu_2024_run.txt)
[nvidia_2024_run.txt](https://github.com/oneapi-src/oneMKL/files/13583720/nvidia_2024_run.txt)

**Comment by lhuot** [comment]
@FMarno there are a lot of tests skipped, I am assuming it is because portFFT does not support the FFT problems being tested, is that correct?

**Comment by FMarno** [comment]
> @FMarno there are a lot of tests skipped, I am assuming it is because portFFT does not support the FFT problems being tested, is that correct?

That is correct, support is still a bit limited. Full details are here: https://github.com/codeplaysoftware/portFFT#supported-configurations

**Comment by lhuot** [comment]
> > @FMarno there are a lot of tests skipped, I am assuming it is because portFFT does not support the FFT problems being tested, is that correct?
> 
> That is correct, support is still a bit limited. Full details are here: https://github.com/codeplaysoftware/portFFT#supported-configurations

I think it would make sense to add an additional note in the documentation of this project about limited support in portFFT.

**Comment by FMarno** [comment]
> > > @FMarno there are a lot of tests skipped, I am assuming it is because portFFT does not support the FFT problems being tested, is that correct?
> > 
> > 
> > That is correct, support is still a bit limited. Full details are here: https://github.com/codeplaysoftware/portFFT#supported-configurations
> 
> I think it would make sense to add an additional note in the documentation of this project about limited support in portFFT.

I've added https://github.com/oneapi-src/oneMKL/pull/416/commits/3e9128ca7186eec59623ffde4f351bae04819ac3 to hopefully make the support clear

**Review by lhuot — APPROVED** [review]

**Pull request merged**